feat(appmesh): add listener timeout to Virtual Nodes

sshver opened the pull request:
Adds listener timeout to Virtual Nodes.

BREAKING CHANGE: `IVirtualNode` no longer has the `addBackends()` method. A backend can be added to `VirtualNode` using the `addBackend()` method which accepts a single `IVirtualService`
* **appmesh**: `IVirtualNode` no longer has the `addListeners()` method. A listener can be added to `VirtualNode` using the `addListener()` method which accepts a single `VirtualNodeListener`
* **appmesh**: `VirtualNode` no longer has a default listener. It is valid to have a `VirtualNode` without any listeners
* **appmesh**: the construction property `listener` of `VirtualNode` has been renamed to `listeners`, and its type changed to an array of listeners
* **appmesh**: the struct `VirtualNodeListener` has been removed. To create Virtual Node listeners, use the static factory methods of the `VirtualNodeListener` class
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c0073e694157042c4cf910d81a38d50ebaf21a8d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c859adee-25ae-443c-bb8a-e82d238c5d3a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 97e1275af04db44655f75fc87a68975aaed158d5
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3b2c90b0-8289-4478-b6bb-bbe924c92097%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (commented):
Thanks for the contribution @sshver !

Before I do a full review, I wanted to ask one thing I noticed. In the `VirtualNodeListener`, there is a `portMapping` property, that has a `protocol` property. I can't help but notice that the 4 values for this `Protocol` enum correspond to the timeout values here.

I assume that you can only set the timeout for a protocol that you've set in `portMapping`, and doing it for a different protocol is either an error, or is silently ignored? If that's the case, I wonder whether we're missing a bigger opportunity here to model this in a richer way. Just off the top of my head, what if `Protocol` was a more sophisticated type instead of just a simple enum, and you could do things like:

```ts
const node = new VirtualNode(this, 'node', {
  listener: {
    portMapping: {
      port: 8080,
      protocol: Protocol.http({
        idleTimeout: Duration.minutes(1),
        requestTimeout: Duration.minutes(2),
      }),
    },
    // ...
```

Let me know what you think of this direction!

dfezzie commented:
> Thanks for the contribution @sshver !
> 
> Before I do a full review, I wanted to ask one thing I noticed. In the `VirtualNodeListener`, there is a `portMapping` property, that has a `protocol` property. I can't help but notice that the 4 values for this `Protocol` enum correspond to the timeout values here.
> 
> I assume that you can only set the timeout for a protocol that you've set in `portMapping`, and doing it for a different protocol is either an error, or is silently ignored? If that's the case, I wonder whether we're missing a bigger opportunity here to model this in a richer way. Just off the top of my head, what if `Protocol` was a more sophisticated type instead of just a simple enum, and you could do things like:
> 
> ```ts
> const node = new VirtualNode(this, 'node', {
>   listener: {
>     portMapping: {
>       port: 8080,
>       protocol: Protocol.http({
>         idleTimeout: Duration.minutes(1),
>         requestTimeout: Duration.minutes(2),
>       }),
>     },
>     // ...
> ```
> 
> Let me know what you think of this direction!

We do use `Protocol` throughout the API with varying context. For example, you can use `Protocol` in the health check portion of virtual nodes, where the `idle` and `perRequest` may not necessarily map. 

How could you distinguish between places where you can use the properties?

skinny85 commented:
Before I answer that @dfezzie , can you answer my original question?

> I assume that you can only set the timeout for a protocol that you've set in `portMapping`, and doing it for a different protocol is either an error, or is silently ignored?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 03e012f001c42ceaf78a6577a3bf54c604d54d5f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=52776692-218c-4b49-9464-da3282a162d0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

sshver commented:
> Before I answer that @dfezzie , can you answer my original question?
> 
> > I assume that you can only set the timeout for a protocol that you've set in `portMapping`, and doing it for a different protocol is either an error, or is silently ignored?

The timeout should be set up for a protocol, the listener uses. Doing it for a different protocol is an error. This was a miss from my end. I have updated the pull request and fixed this error.

skinny85 commented:
Thanks @sshver and @dfezzie . Here's my issue.

So in the current modeling, we have a bunch of completely separate properties (`portMapping`, `healthCheck`, `timeout` ) in `VirtualNodeListener`. Separate properties usually imply they can be set separately - but that is not the case here. When I set `protocol` in `portMapping` to `HTTP`, I suddenly have a constraint that I also have to use `http` for the `timeout` key (I also expect there is a similar constraint with `healthCheck`).

I'm wondering whether we can capture these invariants into a class that makes it obvious at compile-time what those different constraints are. Something like this:

```ts
new appmesh.VirtualNode(this, 'Node', {
  listener: appmesh.Listeners.httpListener({
    timeout: {
      // only HTTP timeout properties are valid here!
     // ...
    },
    healthCheck: {
      // no need to specify Protocol here - we know it's HTTP
});
```

And we would have methods like `tcpListener()`, `grpcListener()`, etc.

Thoughts on this?

dfezzie commented:
@skinny85 I was thinking along these lines as well. I think it is a good pattern for our service to adopt in the CDK as we have multiple instances where we are setting different properties based on the protocol. I implemented this to see how it would work for `Gateway Routes` as I was working on those during this thread and it works well. It is definitely easier to reason about as a user of CDK and makes implementation a bit simpler as well. 

It would require us rewriting a few of the implementations as they stand, but I'm thinking it may be worth the effort

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c520b012dd2b77f35f80760fa6b6a6a22d7393ef
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cdb22510-9963-4d05-a648-d6392e3081dc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b348e4262986d4f51bccbd3909e0115ca7f6fbec
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d7b0e651-a8ca-4274-878a-1a20d21406ef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2fe07eb387f6a91ea854562e051c88cc1398c560
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8cfd1c08-3809-4831-9598-40bbb1e73371%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9d4f03b6cfbbaae645fbcb917d617181bc8f8381
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21bd6cb2-01ad-4ca2-9be7-ceff79c07e25%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+export class HttpNodeListener extends VirtualNodeListener {
+/**
+ * Returns the ListenerTimeoutProperty for HTTP protocol
+ */
```
This comment block is indented weirdly...

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Returns the ListenerTimeoutProperty for HTTP protocol
+ */
+  public static renderTimeout(tm: HttpTimeout): CfnVirtualNode.ListenerTimeoutProperty {
```
Same as above - this should not be part of the public API of this module, this should be a module-private utility function.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Represents the properties required to define a HTTP Listener for a VirtualNode.
+ */
+export class HttpNodeListener extends VirtualNodeListener {
```
I don't think there's any reason to export this class...?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+  /**
+  * Binds the current object when adding Listener to a VirtualNode
+  */
+  public abstract bind(scope: cdk.Construct): CfnVirtualNode.ListenerProperty;
```
Wrap this in a `VirtualNodeListenerConfig` interface with a single property of type `CfnVirtualNode.ListenerProperty`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Represents the properties required to define a HTTP2 Listener for a VirtualNode.
+ */
+export class Http2NodeListener extends VirtualNodeListener {
```
Same here - I don't think there's any reason to export this class.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+        } : undefined,
+      },
+    });
+  }
```
All of this is copy-pasted from `HttpNodeListener`.

Perhaps `HttpNodeListener` and `Http2NodeListener` should share the same `timeout` type?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+  /**
+   * Return Listener for HTTP2 protocol when Listener is added to Virtual Node.
+   */
+  public bind(_scope: cdk.Construct): CfnVirtualNode.ListenerProperty {
+    return {
+      portMapping: {
+        port: this.port,
+        protocol: Protocol.HTTP2,
+      },
+      healthCheck: Http2NodeListener.renderHealthCheck({
+        port: this.port,
+        protocol: Protocol.HTTP2,
+      }, this.healthCheck),
+      timeout: this.timeout ? Http2NodeListener.renderTimeout(this.timeout) : undefined,
+    };
+  }
```
From what I can see, this is duplicated from the `HttpNodeListener` class. We need to have a single class that represents those two listeners, that you can pass either `HTTP` or `HTTP2` as the protocol, to get rid of this duplication.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Represents the properties required to define a GRPC Listener for a VirtualNode.
+ */
+export class GrpcNodeListener extends VirtualNodeListener {
```
Same here - should not be exported.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Returns the ListenerTimeoutProperty for GRPC protocol
+ */
+  public static renderTimeout(tm: GrpcTimeout): CfnVirtualNode.ListenerTimeoutProperty {
```
Same here - this is a utility, not a public API of this class.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Represents the properties required to define a TCP Listener for a VirtualNode.
+ */
+export class TcpNodeListener extends VirtualNodeListener {
```
Same here - should not be exported.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Returns the ListenerTimeoutProperty for TCP protocol
+ */
+  public static renderTimeout(tm: TcpTimeout): CfnVirtualNode.ListenerTimeoutProperty | undefined {
```
Same here - should be a utility, not a public API of this class.

(BTW, there's a ton of duplication in these. You might want to think about DRYing this whole `VirtualNodeListener` hierarchy up)

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -89,7 +528,7 @@ export interface VirtualNodeBaseProps {
    * @default - No listeners
    */
-  readonly listener?: VirtualNodeListener;
+  readonly listeners?: VirtualNodeListener[];
```
Why this change....? Considering there can only be a single Listener according to the validation above, I don't think this change is needed.

sshver replied:
As discussed offline with @dfezzie, we will be accepting the array of listeners for Virtual Node without validation.

skinny85 replied:
Then you need to add a BREAKING CHANGE related to this. See: https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#step-4-commit

skinny85 replied:
Still missing the breaking change in the PR description.

skinny85 reviewed (changes requested):
Nice work @sshver ! Some notes below, mainly I don't like the change from a single `listener` to having many of them in `VirtualNode`, and also the duplication present in the `VirtualNodeListener` hierarchy must be addressed.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b012bbbd38f1140bc5f0ea3d9aa867934262ab3d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21fe36bc-5795-4c59-be7e-101f6b029897%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dfezzie commented on packages/@aws-cdk/aws-appmesh/lib/index.ts:
```diff
@@ -6,3 +6,4 @@ export * from './shared-interfaces';
 export * from './virtual-node';
 export * from './virtual-router';
 export * from './virtual-service';
+export * from './virtual-listener';
```
I think we should rename this to `virtual-node-listener`. No reason to not be specific

dfezzie commented on packages/@aws-cdk/aws-appmesh/lib/virtual-listener.ts:
```diff
@@ -0,0 +1,338 @@
+/**
+ * Returns an HealthCheck for a VirtualNode
+ */
+function renderHealthCheck(pm: PortMapping, hc: HealthCheck | undefined): CfnVirtualNode.HealthCheckProperty | undefined {
```
This can be made a protected function of `VirtualNodeListener` and then you can use the `protocol` and `port` of the class instead of having to pass in the `PortMapping` interface.

dfezzie commented on packages/@aws-cdk/aws-appmesh/lib/virtual-listener.ts:
```diff
@@ -0,0 +1,338 @@
+/**
+ * Returns the ListenerTimeoutProperty for HTTP protocol
+ */
+function renderTimeout(tm: any, pr: Protocol): CfnVirtualNode.ListenerTimeoutProperty {
```
Can this be a protected function of `VirtualNodeListener`? That way you can just reference the class properties and not pass parameters to this function

dfezzie commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -137,75 +143,26 @@ abstract class VirtualNodeBase extends cdk.Resource implements IVirtualNode {
    */
   public addListeners(...listeners: VirtualNodeListener[]) {
-    if (this.listeners.length + listeners.length > 1) {
+    if (listeners.length + this.listeners.length > 1) {
       throw new Error('VirtualNode may have at most one listener');
     }
```
We discussed this offline, but wanted to put it here.

We are going to accept the array of listeners without validation. This is to future proof against when we lift the server side validation for a single listener. We do not want to require customers to upgrade their CDK to use multiple listeners

dfezzie commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -258,7 +215,7 @@ export class VirtualNode extends VirtualNodeBase {
 
     this.addBackends(...props.backends || []);
-    this.addListeners(...props.listener ? [props.listener] : []);
+    this.addListeners(...props.listener ? [props.listener] : [VirtualNodeListener.httpNodeListener()]);
```
We should not default to a listener for virtual nodes. It is valid to create a virtual node without a listener. For example, we may have a service that polls SQS to do some work, and therefore would not need a listener

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8c8c2adf48495c2575638196fa1f2dbf9f93fe86
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=14dc1dfa-6b91-40cc-89c9-6502224b98d9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8373e3c87d5c0c8078e8ba7c9fd7e6641c33474b
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e84421a7-fd58-491d-9d81-f95531492fab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f220debd562ae009bf66c080edb351f27f0601dd
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=247049f6-a9ab-4f0a-9999-a4a51f75a798%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

sshver commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -186,14 +186,20 @@ const node = new VirtualNode(this, 'node', {
       timeout: Duration.seconds(2), // min
       unhealthyThreshold: 2,
     },
+    timeout: {
```
Yes, I did not update the documentation earlier. But this is now updated in the new revision.

sshver commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts:
```diff
@@ -264,12 +264,9 @@ export class AppMeshExtension extends ServiceExtension {
-      },
+      listeners: [appmesh.VirtualNodeListener.httpNodeListener({
+        port: containerextension.trafficPort,
+      })],
```
I have added a function to add the listener to the virtual node based on protocol.

sshver commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+/**
+ * Represents the properties needed to define a Listeners for a VirtualNode
+ */
+export interface VirtualNodeListenerProps {
```
Created `virtual-node-listener` and moved the related item to that file.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ae6ad41637ef8d2c5d83bf21dcdbc0639ccca8d1
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=058b9958-8532-4daa-ad68-b26204e4e968%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts:
```diff
@@ -259,17 +259,33 @@ export class AppMeshExtension extends ServiceExtension {
       throw new Error('You must add a CloudMap namespace to the ECS cluster in order to use the AppMesh extension');
     }
 
+    function addListener(protocol: appmesh.Protocol, port: number) {
```
Please specify the return type explicitly.

skinny85 commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts:
```diff
@@ -259,17 +259,33 @@ export class AppMeshExtension extends ServiceExtension {
     }
 
+    function addListener(protocol: appmesh.Protocol, port: number) {
+      if (protocol === appmesh.Protocol.HTTP2) {
```
Can you use a `switch` instead?

```ts
switch (protocol) {
  case appmesh.Protocol.HTTP2: return appmesh.VirtualNodeListener.http2NodeListener({ port });
  // ...
}
```

This way, the compiler will guarantee that all enum cases are covered.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -88,22 +88,53 @@ export interface PortMapping {
+   *
+   * @default - none
+   */
+  readonly idle?: cdk.Duration;
```
I believe all these types are used only in the `virtual-node-listener.ts` file. Can you move them there please?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,348 @@
+  /**
+   * Returns an HTTP Listener for a VirtualNode
+   */
+  public static httpNodeListener(props: HttpNodeListenerProps = {}): VirtualNodeListener {
```
Same comment here as in https://github.com/aws/aws-cdk/pull/11277#discussion_r517000567. Should we call these `http` / `http2` / `grpc` / etc.?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,348 @@
+  /**
+   * Returns an HTTP2 Listener for a VirtualNode
+   */
+  public static http2NodeListener(props: HttpNodeListenerProps = {}): VirtualNodeListener {
```
Can you move these `*Props` interfaces above the `VirtualNodeListener` class please?

Also, we generally call them `Props` only when they're used for Construct properties. For static methods like these, they should have the suffix 'Options' (so, `HttpNodeListenerOptions`).

sshver replied:
These `*NodeListenerProps` are being used to initialize the constructor of the `*NodeListener` classes as the static method returns `*NodeListener` object and the props are used to initialize the constructor. Wouldn't it be okay to call them `*Props` here ?

skinny85 replied:
No, because `VirtualNodeListener` is not a construct. We use `*Props` only for construct properties.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,348 @@
+  /**
+   * Return Listener for TCP protocol when Listener is added to Virtual Node.
+  */
+  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
```
All of these `bind`s are exactly the same... can you move this implementation up to `VirtualNodeListener`, and use `abstract` properties there like `protocol` to get rid of this duplication?

sshver replied:
The timeout property is not generic for a VirtualNodeListener. i.e. the timeout property for GRPC protocol would be of the type `GrpcTimeout` whereas for TCP protocol it is `TcpTimeout`. The `TcpTimeout` and `GrpcTimeout` are different as they have different properties.
```
export interface GrpcTimeout {

  readonly idle?: cdk.Duration;

  readonly perRequest?: cdk.Duration;
}

export interface TcpTimeout {

  readonly idle?: cdk.Duration;
}
```
Even though the bind are exactly the same, the timeout property would make them different. Therefore `bind` methods was implemented for each `*NodeListener` class.

skinny85 replied:
Then do this:

```ts
// module-private
interface Timeout {
  readonly idle?: cdk.Duration;

  readonly perRequest?: cdk.Duration;
}
```

And use `Timeout` as the type in `VirtualNodeListener`.

Because of TypeScript's structural typing, you can pass instances of `GrpcTimeout` and `TcpTimeout` where `Timeout` is expected.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
   /**
    * Utility method to add Node Listeners for new or existing VirtualNodes
    */
-  addListeners(...listeners: VirtualNodeListener[]): void;
```
I still haven't seen a reply to this 🙂

sshver replied:
Without rest parameter: `addListener(listeners: VirtualNodeListener[])`
```
node.addListeners([appmesh.VirtualNodeListener.tcp({
  port: 80,
  healthCheck: { timeout: cdk.Duration.seconds(3) },
  timeout: { idle: cdk.Duration.seconds(10) },
}),
appmesh.VirtualNodeListener.http2({
  port: 80,
  healthCheck: {},
  timeout: { idle: cdk.Duration.seconds(10) },
})]);
```
With rest parameter: `addListeners(...listener: VirtualNodeListener[])`

```
node.addListeners(appmesh.VirtualNodeListener.tcp({
  port: 80,
  healthCheck: { timeout: cdk.Duration.seconds(3) },
  timeout: { idle: cdk.Duration.seconds(10) },
}),
appmesh.VirtualNodeListener.http2({
  port: 80,
  healthCheck: {},
  timeout: { idle: cdk.Duration.seconds(10) },
}));
```
If I understand correctly, there are not differences in using either one of them except one would accept array of VirtualListenerNode and other would accept any number of VirtualNodeListener type objects. But to introduce less changes to the existing code, I guess we should use the rest parameters or is this change acceptable ?

dfezzie replied:
My question is, should we be adding this method to the `IVirtualNode` class at all? It would not make sense for an imported virtual node to add listeners, or even backends since those would need to be able to mutate the underlying resource.

skinny85 replied:
> My question is, should we be adding this method to the `IVirtualNode` class at all? It would not make sense for an imported virtual node to add listeners, or even backends since those would need to be able to mutate the underlying resource.

In that case, we should move this method to the `VirtualNode` class, instead of having it on `IVirtualNode`.

Can Virtual Nodes today have more than one Listener?

dfezzie replied:
Not today, but similar to other resources we have supporting multiple listeners on the roadmap

skinny85 replied:
In that case, let's remove the `addListeners` method completely, and move the `addListener` from `IVirtualNode` to `VirtualNode`.

skinny85 replied:
We need a breaking change in the PR description for this as well.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.health-check.ts:
```diff
@@ -134,11 +149,14 @@ export = {
+          protocol,
+          path: '/',
+        },
+      })],
```
I gotta say, I don't understand why do we need changes to these tests...?

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -71,15 +70,12 @@ export = {
         });
 
-        const node = mesh.addVirtualNode('test-node', {
+        new appmesh.VirtualNode(stack, 'test-node', {
```
Why this change...?

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -113,19 +273,11 @@ export = {
-        protocol: appmesh.Protocol.TCP,
-      },
-    });
+    appmesh.VirtualNode.fromVirtualNodeName(stack2, 'imported-node', mesh.meshName, node.virtualNodeName);
```
Sorry, why these changes here...?

sshver replied:
[Reference](https://github.com/aws/aws-cdk/pull/10793#discussion_r518413076)
We have moved the `addListener` function from `IVirtualNode` to `VirtualNode`. So we cannot use addListener function to add a listener to an imported node.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:
```diff
@@ -275,7 +272,6 @@ export = {
               'Fn::GetAtt': ['meshACDFE68E', 'MeshName'],
             },
             Spec: {
-              // Specifically: no Listeners and Backends
```
Why remove this...?

skinny85 reviewed (changes requested):
We're getting closer! A few more comments.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 03b4c1f0b38554ebce8850b7f63b4a52c2c3e362
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8a9c237c-566c-42db-908e-c7eaa8ba627d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

sshver commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -113,19 +273,15 @@ export = {
 
     const node = mesh.addVirtualNode('test-node', {
       dnsHostName: 'test.domain.local',
-      listener: {},
```
I have removed the empty listener as we will not have a default listener for a node. It would be valid to have a node without listener and then add it later.

[Reference](https://github.com/aws/aws-cdk/pull/10793#discussion_r515393247)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6ff4b82f01bb762b8677571fc5a8e8ec18357249
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a677479-6421-4b32-9e06-eea66e7794e1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a167ae8f2b2821e9a20ecb1eeed813a37636e0a1
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0b27f6e2-3eeb-41b3-8024-2768b4febda9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
````diff
@@ -181,15 +175,18 @@ const node = new VirtualNode(this, 'node', {
 ```
 
-The listeners property can be left blank and added later with the `node.addListeners()` method. The `healthcheck` property is optional but if specifying a listener, the `portMappings` must contain at least one property.
+The listeners property can be left blank and added later with the `node.addListeners()` method. The `healthcheck` and `timeout` properties are optional but if specifying a listener, the `port` must be added.
````
```suggestion
The `listeners` property can be left blank and added later with the `node.addListener()` method. The `healthcheck` and `timeout` properties are optional but if specifying a listener, the `port` must be added.
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -87,25 +87,6 @@ export interface PortMapping {
-   */
-  readonly healthCheck?: HealthCheck;
-}
-
```
I believe that when https://github.com/aws/aws-cdk/pull/11277 is merged, you can update this PR on top of it, and then remove the `PortMapping` interface completely.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Represents the properties needed to define a Listeners for a VirtualNode
+ */
+interface VirtualNodeListenerOptions {
```
Let's call this `VirtualNodeListenerCommonOptions`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Represent the HTTP Node Listener prorperty
+ */
+export interface HttpNodeListenerOptions extends VirtualNodeListenerOptions {
```
I believe this should be called `HttpVirtualNodeListenerOptions`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Represent the GRPC Node Listener prorperty
+ */
+export interface GrpcNodeListenerOptions extends VirtualNodeListenerOptions {
```
Same here (`GrpcVirtualNodeListenerOptions`).

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Represent the TCP Node Listener prorperty
+ */
+export interface TcpNodeListenerOptions extends VirtualNodeListenerOptions {
```
Same here (`TcpVirtualNodeListenerOptions`).

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+    return new TcpNodeListener(props);
+  }
+
+  /**
+   * Protocol the listener implements
+   */
+  protected abstract protocol: Protocol;
+
+  /**
+   * Port to listen for connections on
+   */
+  protected abstract port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   */
+  protected abstract healthCheck?: HealthCheck;
+
+  /**
+   * Represents Listener Timeout
+   */
+  protected abstract timeout?: HttpTimeout;
```
None of these are needed. We should strive to minimize the surface area of public classes.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Represents the properties required to define a HTTP Listener for a VirtualNode.
+ */
+class HttpNodeListener extends VirtualNodeListener {
```
There's no need for these subclasses. They don't have any logic inside of them, except copy-pasted defaulting to port 8080. Here's how to get rid of all this.

```ts
/**
 *  Defines listener for a VirtualNode
 */
export abstract class VirtualNodeListener {
  /**
   * Returns an HTTP Listener for a VirtualNode
   */
  public static http(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.HTTP, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Returns an HTTP2 Listener for a VirtualNode
   */
  public static http2(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.HTTP2, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Returns an GRPC Listener for a VirtualNode
   */
  public static grpc(props: GrpcNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.GRPC, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Binds the current object when adding Listener to a VirtualNode
   */
  public abstract bind(scope: cdk.Construct): VirtualNodeListenerConfig;
}

// module-private
class VirtualNodeListenerImpl extends VirtualNodeListener {
  constructor(private readonly protocol: Protocol, 
    private readonly healthCheck: HealthCheck | undefined,
    private readonly timeout: HttpTimeout | undefined,
    private readonly port: number = 8080) {
  }

  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
    return {
      listener: {
        portMapping: {
          port: this.port,
          protocol: this.protocol,
        },
        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
        timeout: this.timeout ? this.renderTimeout(this.timeout) : undefined,
      },
    };
  }

  private renderHealthCheck(hc: HealthCheck): CfnVirtualNode.HealthCheckProperty | undefined {
    if (hc === undefined) { return undefined; }

    if (hc.protocol === Protocol.TCP && hc.path) {
      throw new Error('The path property cannot be set with Protocol.TCP');
    }

    if (hc.protocol === Protocol.GRPC && hc.path) {
      throw new Error('The path property cannot be set with Protocol.GRPC');
    }

    const healthCheck: CfnVirtualNode.HealthCheckProperty = {
      healthyThreshold: hc.healthyThreshold || 2,
      intervalMillis: (hc.interval || cdk.Duration.seconds(5)).toMilliseconds(), // min
      path: hc.path || (hc.protocol === Protocol.HTTP ? '/' : undefined),
      port: hc.port || this.port,
      protocol: hc.protocol || this.protocol,
      timeoutMillis: (hc.timeout || cdk.Duration.seconds(2)).toMilliseconds(),
      unhealthyThreshold: hc.unhealthyThreshold || 2,
    };

    validateHealthChecks(healthCheck);

    return healthCheck;
  }

  private renderTimeout(timeout: HttpTimeout): CfnVirtualNode.ListenerTimeoutProperty {
    return ({
      [this.protocol]: {
        idle: timeout?.idle !== undefined ? {
          unit: 'ms',
          value: timeout?.idle.toMilliseconds(),
        } : undefined,
        perRequest: timeout?.perRequest !== undefined ? {
          unit: 'ms',
          value: timeout?.perRequest.toMilliseconds(),
        } : undefined,
      },
    });
  }
}

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -263,6 +196,26 @@ export class VirtualNode extends VirtualNodeBase {
+  /**
+   * Utility method to add an inbound listener for this VirtualNode
+   */
+  public addListener(...listeners: VirtualNodeListener[]) {
```
`addListener()` should take a single `VirtualNodeListener`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -122,69 +113,11 @@ abstract class VirtualNodeBase extends cdk.Resource implements IVirtualNode {
-  validateHealthChecks(healthCheck);
-
-  return healthCheck;
+  protected readonly listeners = new Array<VirtualNodeListenerConfig>();
```
Why is this here? This should be in `VirtualNode`, right?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -122,69 +113,11 @@ abstract class VirtualNodeBase extends cdk.Resource implements IVirtualNode {
   public abstract readonly virtualNodeArn: string;
 
   protected readonly backends = new Array<CfnVirtualNode.BackendProperty>();
```
This field should be moved to `VirtualNode`, it's not used in this class.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -263,6 +196,26 @@ export class VirtualNode extends VirtualNodeBase {
+  /**
+   * Add a Virtual Services that this node is expected to send outbound traffic to
+   */
+  public addBackends(...props: IVirtualService[]) {
```
Let's make `addBackends()` consistent with `addListener()` (call it `addBackend()`, make it take a single `IVirtualService`).

We will need a breaking change in the PR description for this too.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -98,6 +94,170 @@ export = {
 
+        test.done();
+      },
+    },
+    'when a listener is added with timeout': {
```
There should be an empty line between tests on the same level:

```suggestion
    },

    'when a listener is added with timeout': {
```

Same comment for all new tests added.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -98,6 +94,170 @@ export = {
+        });
+
+        // THEN
+        expect(stack).to(
+          haveResourceLike('AWS::AppMesh::VirtualNode', {
```
These should be on the same line, because otherwise the indentation is weird:

```suggestion
        expect(stack).to(haveResourceLike('AWS::AppMesh::VirtualNode', {
```

Same comment for all other added tests in this PR.

skinny85 reviewed (changes requested):
Looks good! We're almost there, just needs some finishing touches.

Also make sure to update the PR description, so that int contains all of the breaking changes made in this PR (which is _a lot_): https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#step-4-commit

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0d70cd74bb30d3bc9a5883db49da609b4cc82958
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=73cd0dfc-eded-48be-9439-3ce656c9277f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b67d48ebc81a26af1514404d274bc7b596bf88a1
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ac92b398-f362-4ffd-903e-aec700648ba6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a2138cd0356e13640933e49a093ff6311241fa80
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ca22d774-0741-49e7-9ccb-83bf805d9204%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (approved):
Looks fantastic @sshver !

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: caf3e7cf9ea45f36ebe95ca07dfe605080532c10
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=20f1753e-6866-404a-80ea-3218442735e9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d35705f855fc3c967a370126dc229fd236f5e1a0
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=39dee008-d8eb-459e-8e3c-34bd925c7e3e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: be69829d37dce46151b409b121f8fa0209d0b4f1
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f0bc16b5-841f-4a0e-9a54-8d859f3e1a73%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d8b049ccc31cb8af523d4018ab7a03b32f232d96
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=573e5c6d-7995-4612-9d3a-f1454c470ba7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cf15f6e495b5abc8a94a37dc1778e694f79b782a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4ebff5a0-ad9a-43b2-b97c-2ba2953fc96a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
